src/action.c

```diff
@@ -267,9 +335,70 @@ actions_run(struct view *activator, struct server *server,
 			 * adding a new action without installing a handler here.
 			 */
 			wlr_log(WLR_ERROR,
-				"Not executing invalid action (%u) with arg %s"
-				" This is a BUG. Please report.",
-				action->type, action->arg);
+				"Not executing invalid action (%u)"
+				" This is a BUG. Please report.", action->type);
 		}
 	}
 }
+
+
+void
+action_add_str_arg(struct action *action, char *name, const char *value)
+{
+	assert(value && "Tried to add NULL action string argument");
+	struct action_arg_str *arg = calloc(1, sizeof(*arg));
+	arg->base.type = LAB_ACTION_ARG_STR;
+	if (name) {
+		arg->base.name = strdup(name);
+	}
+	arg->value = strdup(value);
+	wl_list_insert(action->args.prev, &arg->base.link);
+}
+
+void
+action_add_int_arg(struct action *action, char *name, int value)
+{
+	struct action_arg_int *arg = calloc(1, sizeof(*arg));
+	arg->base.type = LAB_ACTION_ARG_INT;
+	if (name) {
+		arg->base.name = strdup(name);
+	}
+	arg->value = value;
+	wl_list_insert(action->args.prev, &arg->base.link);
+}
+
+void
+action_add_double_arg(struct action *action, char *name, double value)
+{
+	struct action_arg_double *arg = calloc(1, sizeof(*arg));
+	arg->base.type = LAB_ACTION_ARG_DOUBLE;
+	if (name) {
+		arg->base.name = strdup(name);
+	}
+	arg->value = value;
+	wl_list_insert(action->args.prev, &arg->base.link);
+}
+
+void
+action_add_bool_arg(struct action *action, char *name, bool value)
+{
+	struct action_arg_bool *arg = calloc(1, sizeof(*arg));
+	arg->base.type = LAB_ACTION_ARG_BOOL;
+	if (name) {
+		arg->base.name = strdup(name);
+	}
+	arg->value = value;
+	wl_list_insert(action->args.prev, &arg->base.link);
+}
+
+struct wlr_box *
+action_add_box_arg(struct action *action, char *name)
+{
+	struct action_arg_box *arg = calloc(1, sizeof(*arg));
+	arg->base.type = LAB_ACTION_ARG_BOX;
+	if (name) {
+		arg->base.name = strdup(name);
+	}
+	wl_list_insert(action->args.prev, &arg->base.link);
+	return &arg->value;
+}
```

Comment:
Only `action_add_str_arg()` is used in this PR, the others are examples of how this action_arg framework could be used in the future